### PR TITLE
[SPARK-7383][ML] Feature Parity in PySpark for ml.features

### PR DESCRIPTION
**Pull request by brkyvz**
Implemented python wrappers for Scala functions that don't exist in `ml.features`

### CODE REVIEW

File: python/pyspark/ml/feature.py [hunk 1]
```diff
@@ -16,61 +16,73 @@
...
-    >>> tokenizer.transform(df, {tokenizer.outputCol: "words"}).head()
-    Row(text=u'a b c', words=[u'a', u'b', u'c'])
-    >>> tokenizer.transform(df).head()
-    Row(text=u'a b c', tokens=[u'a', u'b', u'c'])
-    >>> # Must use keyword arguments to specify params.
-    >>> tokenizer.setParams("text")
-    Traceback (most recent call last):
-        ...
-    TypeError: Method setParams forces keyword arguments.
+    >>> df = sc.parallelize([Row(values=0.5)]).toDF()
```
Comment (mengxr) [inline python/pyspark/ml/feature.py]: minor: I'm not sure which one is the recommended approach to create a DataFrame. @rxin


`df = sc.parallelize([Row(values=0.5)]).toDF()`

vs.

`df = sqlContext.createDataFrame([(0.5,)], ["values"])`  # don't need to import Row

File: python/pyspark/ml/feature.py [hunk 2]
```diff
@@ -16,61 +16,73 @@
...
+    >>> Binarizer = Binarizer(threshold=1.0, inputCol="values", outputCol="features")
```
Comment (mengxr) [inline python/pyspark/ml/feature.py]: the first `Binarizer` -> `binarizer`

File: python/pyspark/ml/feature.py [hunk 3]
```diff
@@ -113,6 +125,419 @@ def setParams(self, numFeatures=1 << 18, inputCol=None, outputCol=None):
...
+@inherit_doc
+class Normalizer(JavaTransformer, HasInputCol, HasOutputCol):
+    """
+     Normalize a vector to have unit norm using the given p-norm.
+
+    >>> from pyspark.mllib.linalg import Vectors
+    >>> from pyspark.sql import Row
+    >>> svec = Vectors.sparse(4, {1: 4.0, 3: 3.0})
+    >>> df = sc.parallelize([Row(dense=Vectors.dense([3.0, -4.0]), sparse=svec)]).toDF()
+    >>> Normalizer = Normalizer(p=2.0, inputCol="dense", outputCol="features")
```
Comment (mengxr) [inline python/pyspark/ml/feature.py]: the first `Normalizer` -> `normalizer`

File: python/pyspark/ml/feature.py [hunk 4]
```diff
@@ -113,6 +125,419 @@ def setParams(self, numFeatures=1 << 18, inputCol=None, outputCol=None):
...
+    at most a single one-value. By default, the binary vector has an element for each category, so
+    with 5 categories, an input value of 2.0 would map to an output vector of
+    (0.0, 0.0, 1.0, 0.0, 0.0). If includeFirst is set to false, the first category is omitted, so
+    the output vector for the previous example would be (0.0, 1.0, 0.0, 0.0) and an input value
+    of 0.0 would map to a vector of all zeros. Including the first category makes the vector columns
+    linearly dependent because they sum up to one.
+
+    TODO: This method requires the use of StringIndexer first. Decouple them.
+
+    >>> StringIndexer = StringIndexer(inputCol="label", outputCol="indexed")
```
Comment (mengxr) [inline python/pyspark/ml/feature.py]: `StringIndexer` -> `stringIndexer`

File: python/pyspark/ml/feature.py [hunk 5]
```diff
@@ -113,6 +125,419 @@ def setParams(self, numFeatures=1 << 18, inputCol=None, outputCol=None):
...
+@inherit_doc
+class StandardScaler(JavaEstimator, HasInputCol, HasOutputCol):
+    """
+    Standardizes features by removing the mean and scaling to unit variance using column summary
+    statistics on the samples in the training set.
+
+    >>> from pyspark.mllib.linalg import Vectors
+    >>> from pyspark.sql import Row
+    >>> df = sc.parallelize([Row(a=Vectors.dense([0.0])), Row(a=Vectors.dense([2.0]))]).toDF()
+    >>> StandardScaler = StandardScaler(inputCol="a", outputCol="scaled")
```
Comment (mengxr) [inline python/pyspark/ml/feature.py]: the first `StandardScaler` -> `standardScaler`

File: python/pyspark/ml/feature.py [hunk 6]
```diff
@@ -113,6 +125,419 @@ def setParams(self, numFeatures=1 << 18, inputCol=None, outputCol=None):
...
+
+@inherit_doc
+class StringIndexer(JavaEstimator, HasInputCol, HasOutputCol):
+    """
+    A label indexer that maps a string column of labels to an ML column of label indices.
+    If the input column is numeric, we cast it to string and index the string values.
+    The indices are in [0, numLabels), ordered by label frequencies.
+    So the most frequent label gets index 0.
+
+    >>> StringIndexer = StringIndexer(inputCol="label", outputCol="indexed")
```
Comment (mengxr) [inline python/pyspark/ml/feature.py]: `StringIndexer` -> `stringIndexer`

**Comment by mengxr** [comment]
@brkyvz Thanks for working on this! It looks good except the variable naming in the doctests. It seems that `RegexTokenizer` is missing from the list. Could you add it as well?

File: python/pyspark/ml/feature.py (hunk 1 above)
Comment (rxin) [inline python/pyspark/ml/feature.py]: I prefer the 2nd approach

**Comment by mengxr** [comment]
LGTM pending Jenkins.

**Comment by mengxr** [comment]
Merged into master and branch-1.4. Thanks!

**Pull request closed without merging**